### PR TITLE
ui: migrate canvas to nodes

### PR DESCRIPTION
**Pull request by psychedelicious**
# migrate canvas to nodes

many changes across the UI and some changes to nodes backend

## nodes

- image upload route accepts an image type. this way we can upload the intermediate canvas images (eg the canvas and its mask which are used for img2img) as intermediates
- make seed optional w/ default factory to generate random seeds
- add infill nodes (that's as far as I've gotten with porting the generator Inpaint class over to nodes)
- add `ColorField` (supporting solid infill method)
- add `ImageToLatentsInvocation` (VAE Encode)
- improve `choose_model` logic and messaging
- fix the types in `inpaint.py` in an effort to make sense of it

## ui

well there's a ton. major points: 

- `redux-persist` --> `redux-remember`. modern, maintained, but not very popular for whatever reason. lovely code.
- use RTK `createListenerMiddleware` to handle coordinating the more complex workflows for canvas
- migrate canvas over to nodes (seam paint regression in #3361, missing inpaint width/height params - once canvas backend is fully nodified i will fix this up)
- refactor param components (no more accordions - self-contained modular collapsibles, paves the way for an eventual customizable UI setup)
- refactor tab layouts
- gallery infinite scroll w/ virtualized list (windowing) for performance
- scrollbars! (ya they need tweaking but thats OK for now)
- progress image UI improvements
- many many many bug fixes

one major regression. the responsive design was impossible to keep intact as i went through all these substantial changes. once things settle a bit (after we have controlnet UI figured out basically) i will work on it.

edit: in-canvas progress images are tricky and not yet implemented

### CODE REVIEW

**Comment by psychedelicious** [comment]
i see the metadata test is failing. i'll fix that tomorrow.

File: invokeai/app/api/routers/images.py [hunk 1]
```diff
@@ -83,7 +83,7 @@ async def get_thumbnail(
     status_code=201,
 )
 async def upload_image(
-    file: UploadFile, request: Request, response: Response
+    file: UploadFile, image_type: ImageType, request: Request, response: Response
```
Comment (Kyle0654) [inline invokeai/app/api/routers/images.py:86]: Allowing user to specify type is very risky.

File: invokeai/app/invocations/generate.py [hunk 2]
```diff
@@ -44,15 +47,13 @@ class TextToImageInvocation(BaseInvocation, SDImageInvocation):
...
     # fmt: off
     prompt: Optional[str] = Field(description="The prompt to generate an image from")
-    seed:        int = Field(default=-1,ge=-1, le=np.iinfo(np.uint32).max, description="The seed to use (-1 for a random seed)", )
-    steps:       int = Field(default=10, gt=0, description="The number of steps to use to generate the image")
+    seed:   Optional[int] = Field(ge=0, le=SEED_MAX, description="The seed to use (omit for random)", default_factory=get_random_seed)
+    steps:       int = Field(default=30, gt=0, description="The number of steps to use to generate the image")
     width:       int = Field(default=512, multiple_of=8, gt=0, description="The width of the resulting image", )
     height:      int = Field(default=512, multiple_of=8, gt=0, description="The height of the resulting image", )
-    cfg_scale: float = Field(default=7.5, gt=0, description="The Classifier-Free Guidance, higher values may result in a result closer to the prompt", )
+    cfg_scale: float = Field(default=7.5, ge=1, description="The Classifier-Free Guidance, higher values may result in a result closer to the prompt", )
```
Comment (Kyle0654) [inline invokeai/app/invocations/generate.py:54]: Is this accurate?

File: invokeai/app/invocations/infill.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    si = Image.fromarray(st, mode="RGBA")
+
+    return si
+
+
+class InfillColorInvocation(BaseInvocation):
+    """Infills transparent areas of an image with a solid color"""
+
+    type: Literal["infill_rgba"] = "infill_rgba"
+    image: ImageField = Field(default=None, description="The image to infill")
```
Comment (Kyle0654) [inline invokeai/app/invocations/infill.py:128]: This should have Optional, right?

File: invokeai/app/invocations/latent.py [hunk 4]
```diff
@@ -102,17 +105,13 @@ def get_noise(width:int, height:int, device:torch.device, seed:int = 0, latent_c
...
-
-
 class NoiseInvocation(BaseInvocation):
     """Generates latent noise."""
 
     type: Literal["noise"] = "noise"
 
     # Inputs
-    seed:        int = Field(ge=0, le=np.iinfo(np.uint32).max, description="The seed to use", default_factory=random_seed)
+    seed: Optional[int] = Field(ge=0, le=SEED_MAX, description="The seed to use", default_factory=get_random_seed)
```
Comment (Kyle0654) [inline invokeai/app/invocations/latent.py:114]: I don't think this should be Optional. The default_factory should be used to generate a value if one isn't provided at creation.

**Review by Kyle0654 — APPROVED** [review]
Some changes probably needed, but don't want my review to hold you up.

File: invokeai/app/invocations/generate.py (hunk 2 above)
Comment (psychedelicious) [reply invokeai/app/invocations/generate.py:54]: yes

File: invokeai/app/api/routers/images.py (hunk 1 above)
Comment (psychedelicious) [reply invokeai/app/api/routers/images.py:86]: think the options would be then:
- just make the intermediates `uploads`, but then that kinda defeats the purpose
- add another route for uploading intermediates

any preference or other ideas? the use-case is putting canvas init/mask images in the `intermediates` folder

File: invokeai/app/invocations/infill.py (hunk 3 above)
Comment (psychedelicious) [reply invokeai/app/invocations/infill.py:128]: guess so - i don't think i understand Optional in the context of nodes

Comment (psychedelicious) [reply invokeai/app/invocations/infill.py:128]: there are other places where ImageFields are not Optional, like in the `image.py` nodes - IIRC its to facilitate auto-linking in the cli?

File: invokeai/app/invocations/latent.py (hunk 4 above)
Comment (psychedelicious) [reply invokeai/app/invocations/latent.py]: fixed all Optionals per advice

File: invokeai/app/api/routers/images.py (hunk 1 above)
Comment (psychedelicious) [reply invokeai/app/api/routers/images.py:86]: let's use this issue #3385 to figure out a better way to handle this

**Review by blessedcoolant — APPROVED** [review]

**Pull request merged**